### PR TITLE
Make alert widget not access all flash data (Fixes #250)

### LINKED ISSUE #250: Alert widget marks all flash data for deletion
### What steps will reproduce the problem?

1. Init a new Yii2 project using "yii2-app-basic" template
2. In any route / action that renders a view (or layout) containing the Alert widget, call `Yii::$app->session->setFlash('somekey', 'somevalue')` **before** `$this->render('view');`
3. Open the route from step 2 which renders a view / layout containing the Alert widget
4. Go to another route / action and call `var_dump(Yii::$app->session->getFlash('somekey'));`

### What's expected?

Browser should print `string(9) "somevalue"`

(to be precise, my flash data should be still available because I didn't access it / Alert widget should not touch my flash data)

### What do you get instead?

Browser prints `NULL`

(= my flash data was deleted)

### Additional info

| Q                | A
| ---------------- | ---
| Yii vesion       | 2.0.43
| PHP version      | 7.4.3, 8.0
| Operating system | Ubuntu 20.04, Alpine Linux 3.14

For step 1, I created a new project by calling `composer create-project --prefer-dist yiisoft/yii2-app-basic basic`

For step 2, I extended the SiteController:

```
public function actionFlashTest() {
    Yii::$app->session->setFlash('somekey', 'somevalue');

    return $this->render('about');
}

public function actionFlashTest2() {
    var_dump(Yii::$app->session->getFlash('somekey'));
    die();
}
```

Then I called `./yii serve`.

For step 3, I went to `http://localhost:8080/index.php?r=site%2Fflash-test`

For step 4, I went to `http://localhost:8080/index.php?r=site%2Fflash-test2`

I've digged into the code, reason for this behavior seems to be that the Alert widget calls on line 54:

https://github.com/yiisoft/yii2-app-basic/blob/4551a05d8892cf370ff7d2fff96e81cbb86dd081/widgets/Alert.php#L54

`Yii::$app->session->getAllFlashes();`

But this call marks **all** flash data for deletion on the next request, also flash data other than error, success, info ...

https://github.com/yiisoft/yii2/blob/c21322f0c60ecd66da07b461469f62979141a828/framework/web/Session.php#L847

My fix (barely tested since my project is in an early stage):

// widgets/Alert.php

```
public function run()
{
    $session = Yii::$app->session;
    $appendClass = isset($this->options['class']) ? ' ' . $this->options['class'] : '';

    foreach ($this->alertTypes as $type) {

        $flash = Yii::$app->session->getFlash($type);
        
        foreach ((array) $flash as $i => $message) {
            echo \yii\bootstrap4\Alert::widget([
                'body' => $message,
                'closeButton' => $this->closeButton,
                'options' => array_merge($this->options, [
                    'id' => $this->getId() . '-' . $type . '-' . $i,
                    'class' => $this->alertTypes[$type] . $appendClass,
                ]),
            ]);
        }

        $session->removeFlash($type);
    }
}
```

If someone confirms that this is a bug, I would try to send a pull request (would be my first one ever :-) )

### PR DESCRIPTION
**Pull request by tobias-dietz**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #250

### CODE REVIEW

File: widgets/Alert.php [hunk 1]
```diff
@@ -51,21 +51,18 @@ class Alert extends \yii\bootstrap4\Widget
...
+            $flash = $session->getFlash($type);
 
             foreach ((array) $flash as $i => $message) {
                 echo \yii\bootstrap4\Alert::widget([
                     'body' => $message,
                     'closeButton' => $this->closeButton,
                     'options' => array_merge($this->options, [
-                        'id' => $this->getId() . '-' . $type . '-' . $i,
-                        'class' => $this->alertTypes[$type] . $appendClass,
+                        'id' => $this->getId().'-'.$type.'-'.$i,
```
Comment (bizley) [inline widgets/Alert.php:64]: Please keep the spaces, this is not Symfony 😀

Comment (tobias-dietz) [reply widgets/Alert.php:64]: OK I'll remember that in future :smile: Do I have to push again in the branch I sent the pull request with the added spaces? (as said before, I'm new to GitHub and all the workflows and I don't want to mess up everything, sorry for that newbie questions :laughing: )

Comment (bizley) [reply widgets/Alert.php:64]: Just push it to the same branch ;)

**Comment by tobias-dietz** [comment]
Found the missing spaces and added them :smile: I've also added a unit test. The tests may seem verbose but it felt nessecary since the alert widget has a nested foreach loop and can process strings as well as arrays for the differnt message types. If it's too verbose, I would suggest to remove all tests except for the last three ones. 

The last test tests the fix for this bug and should fail with the current alert widget implementation in master but not with the fixed implementation.

**Comment by bizley** [comment]
Wow, impressive, not everyday we can see so robust testing. One tiny thing - it should be PHPUnit tests strictly, not Codeception so please change the extended test case class and implementation details there (but it should be pretty straight forward if I'm not mistaken).
Also please add CHANGELOG line. Thank you!

**Comment by tobias-dietz** [comment]
Thank you for your nice feedback :smile: I've pushed a new commit, the unit test looks now the same as the existing ones for the models (`tests\unit\models\LoginFormTest` etc.). 
However, I have the strong feeling that I did not understand exactly what you meant because those `expect()` are looking like they only wrap  Codeception methods. 

I tried to extend `PHPUnit\Framework\TestCase` but then I receive a lot of errors like `[yii\base\ErrorException] Attempt to read property "session" on null` (which wasn't that surprising to be honest).

Then I found https://github.com/yiisoft/yii2/blob/master/tests/TestCase.php which I think is what you meant? But it seems that this class is only part of `yiisoft/yii2` and not directly usable in `yiisoft/yii2-app-basic` (at least, PhpStorm doesn't offer that class on auto complete)? If I'm wrong (what is not that unlikely), it would be nice if you could point on additional resources or an example. I followed the TESTING section of the README file in `yiisoft/yii2-app-basic` so far.

Regarding the CHANGELOG: I couldn't find a CHANGELOG in `yiisoft/yii2-app-basic`.

**Comment by bizley** [comment]
Oh, right, no changelog here, weird... Anyway, it looks like I went full-auto here and treated it just like Yii 2 core while it's not the case and the few tests here are indeed Codeception ones - could you please revert the changes I've told you to do and bring back the ones you made before? I'm very very sorry for that extra work 😭

**Comment by tobias-dietz** [comment]
No problem :smile: The commit was broken either way (forgot to clean up my experiments with PHPUnit TestCase). Please have a look again. Basically, they are my original tests, only now with the Codeception `expect()` functions (instead of accessing `$this->tester`) that are also used in the other tests in yii2-app-basic (to maintain visual consistency with the other tests).

**Review by bizley — APPROVED** [review]

**Review by machour — APPROVED** [review]
Nice PR, thank you for your contribution.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thanks!

**Comment by samdark** [comment]
Would you please apply the same to advanced app template?

**Pull request merged**